drivers/sense/stream/argus/main.c

```diff
@@ -0,0 +1,246 @@
+ */
+#include <math.h>
+#include <sys/types.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/shell/shell.h>
+
+#include <zros/perf_duration.h>
+
+#include <synapse_topic_list.h>
+
+#include <zros/private/zros_node_struct.h>
+#include <zros/private/zros_pub_struct.h>
+#include <zros/private/zros_sub_struct.h>
+#include <zros/zros_node.h>
+#include <zros/zros_pub.h>
+#include <zros/zros_sub.h>
+
+#include <../drivers/sensor/broadcom/afbr_s50/afbr_s50_decoder.h>
+
+LOG_MODULE_REGISTER(sense_argus_stream, CONFIG_ZROS_SENSE_STREAM_ARGUS_LOG_LEVEL);
+
+/** Borrowed from zephyr/dsp/util.h as it was erroring out due to a missing
+ * zdsp backend. Should sort that out.
+ */
+#define Z_SHIFT_Q31_TO_F32(src, m) ((float32_t)(((int64_t)src) << m) / (float32_t)(1U << 31))
+
+/** Borrowed from zephyr/dsp/util.h as it was erroring out due to a missing
+ * zdsp backend. Should sort that out.
+ */
+#define Z_SHIFT_Q31_TO_F32(src, m) ((float32_t)(((int64_t)src) << m) / (float32_t)(1U << 31))
+
+#define ARGUS_ALIAS(i) DT_ALIAS(_CONCAT(argus_stream_,i))
+
+static struct context *get_context_from_iodev(const struct rtio_iodev *iodev);
+
+struct context {
+	const char *name;
+	struct zros_node node;
+	struct zros_pub pub;
+	struct zros_topic *pub_topic;
+	bool running;
+	synapse_pb_ArgusResults data;
+	struct {
+		struct rtio *ctx;
+		struct rtio_iodev *iodev;
+	} stream;
+};
+
+static inline const struct sensor_decoder_api *get_decoder(struct rtio_iodev *iodev)
+{
+	struct sensor_read_config *read_config = (struct sensor_read_config *)iodev->data;
+	const struct device *sensor = read_config->sensor;
+	const struct sensor_decoder_api *decoder;
+
+	if (sensor_get_decoder(sensor, &decoder) < 0) {
+		return NULL;
+	}
+	return decoder;
+}
+
+static void argus_publish(struct context *ctx, argus_results_t *payload)
+{
+	stamp_msg(&ctx->data.stamp, k_uptime_ticks());
+
+	ctx->data.status = payload->Status;
+
+	ctx->data.has_bin = true;
+	ctx->data.bin.range = Z_SHIFT_Q31_TO_F32(payload->Bin.Range, 9);
+	ctx->data.bin.amplitude = Z_SHIFT_Q31_TO_F32(payload->Bin.Amplitude, 12);
+	ctx->data.bin.signal_quality = payload->Bin.SignalQuality;
+
+	ctx->data.has_frame = true;
+	ctx->data.frame.integration_time = payload->Frame.IntegrationTime;
+	ctx->data.frame.px_en_mask = payload->Frame.PxEnMask;
+	ctx->data.frame.ch_en_mask = payload->Frame.ChEnMask;
+	ctx->data.frame.state = payload->Frame.State;
+	ctx->data.frame.analog_integration_depth = Z_SHIFT_Q31_TO_F32(payload->Frame.AnalogIntegrationDepth, 10);
+	ctx->data.frame.digital_integration_depth = payload->Frame.DigitalIntegrationDepth;
+	ctx->data.frame.output_power = Z_SHIFT_Q31_TO_F32(payload->Frame.OutputPower, 12);
+	ctx->data.frame.bias_current = payload->Frame.BiasCurrent;
+	ctx->data.frame.pixel_gain = payload->Frame.PixelGain;
+	ctx->data.frame.pll_offset = payload->Frame.PllOffset;
+	ctx->data.frame.pll_ctrl_cur = payload->Frame.PllCtrlCur;
```

Comment:
Maybe we consider this as our new "debug level"?

Reply:
Do you mean putting under `CONFIG_ZROS_SENSE_STREAM_ARGUS_VERBOSE`?

Reply:
Was more thinking along the lines of distance, pointcloud and debug output modes. Debug being where it dumps EVERYTHING, distance is just the bin with status, pointcloud is the pixels with status, and debug once again is all of the above including aux and frame.